File: app/src/core/utility/bcryptTool.test.ts
import { describe, expect, it } from 'vitest'
import { BcryptHasher, BcryptVerifier } from './bcryptTool'

describe('BcryptHasher', () => {
  const hasher = new BcryptHasher()
  const verifier = new BcryptVerifier()

  it('produces a well-formed bcrypt hash string', () => {
    const result = hasher.execute({ plaintext: 'hunter2', logRounds: 4 })
    // $<version>$<cost>$<22-char salt><31-char digest>
    expect(result.hash).toMatch(/^\$2[aby]?\$\d{2}\$[./A-Za-z0-9]{53}$/)
    // bcryptjs emits the $2b$ minor-version marker, not Dart bcrypt's $2a$ --
    // both verify identically, see the class doc in bcryptTool.ts.
    expect(result.hash.startsWith('$2b$04$')).toBe(true)
  })

  it('generates a different hash each time for the same plaintext (random salt)', () => {
    const a = hasher.execute({ plaintext: 'hunter2', logRounds: 4 })
    const b = hasher.execute({ plaintext: 'hunter2', logRounds: 4 })
    expect(a.hash).not.toBe(b.hash)
  })

  it('rejects out-of-range log-rounds', () => {
    expect(() => hasher.execute({ plaintext: 'hunter2', logRounds: 1 })).toThrow()
    expect(() => hasher.execute({ plaintext: 'hunter2', logRounds: 99 })).toThrow()
  })

  describe('BcryptVerifier', () => {
    it('accepts the correct plaintext against a freshly generated hash', () => {
      const hash = hasher.execute({ plaintext: 'correct horse battery staple', logRounds: 4 })
      const result = verifier.execute({ plaintext: 'correct horse battery staple', hash: hash.hash })
      expect(result.matches).toBe(true)
    })

    it('rejects the wrong plaintext against a valid hash', () => {
      const hash = hasher.execute({ plaintext: 'correct horse battery staple', logRounds: 4 })
      const result = verifier.execute({ plaintext: 'wrong password', hash: hash.hash })
      expect(result.matches).toBe(false)
    })

    it('verifies against a pre-recorded fixture hash (fixed-salt regression vector)', () => {
      // bcrypt embeds a random salt in every hash, so there is no
      // cross-implementation "hash of 'abc'" constant the way there is for
      // MD5/SHA-256. This fixture was generated once with Dart's `bcrypt`
      // package (hence the $2a$ marker) and is frozen here so the
      // verifier's decoding of an externally-produced hash string is
      // exercised, not just a live hash-then-immediately-verify round trip.
      // bcryptjs verifies $2a$/$2b$/$2y$ hashes interchangeably.
      const fixtureHash = '$2a$04$y4tiGeWwdpFlQhdt/zqxf.ZbJzRxB2WjyEZXYjWxuSEEsplrY8JQ2'

      const correct = verifier.execute({ plaintext: 'correct horse battery staple', hash: fixtureHash })
      expect(correct.matches).toBe(true)

      const wrong = verifier.execute({ plaintext: 'wrong password', hash: fixtureHash })
      expect(wrong.matches).toBe(false)
    })

    it('returns false (not a thrown error) for a malformed hash string', () => {
      const result = verifier.execute({ plaintext: 'hunter2', hash: 'not-a-bcrypt-hash' })
      expect(result.matches).toBe(false)
    })
  })
})

File: archive/flutter-app/test/core/utility/bcrypt_tool_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:infrakit_studio/core/utility/bcrypt_tool.dart';

void main() {
  const hasher = BcryptHasher();
  const verifier = BcryptVerifier();

  group('BcryptHasher', () {
    test('produces a well-formed bcrypt hash string', () {
      final result = hasher.execute(const BcryptHashInput(plaintext: 'hunter2', logRounds: 4));
      // $<version>$<cost>$<22-char salt><31-char digest>
      expect(result.hash, matches(RegExp(r'^\$2[aby]?\$\d{2}\$[./A-Za-z0-9]{53}$')));
      expect(result.hash, startsWith(r'$2a$04$'));
    });

    test('generates a different hash each time for the same plaintext (random salt)', () {
      final a = hasher.execute(const BcryptHashInput(plaintext: 'hunter2', logRounds: 4));
      final b = hasher.execute(const BcryptHashInput(plaintext: 'hunter2', logRounds: 4));
      expect(a.hash, isNot(b.hash));
    });

    test('rejects out-of-range log-rounds', () {
      expect(
        () => hasher.execute(const BcryptHashInput(plaintext: 'hunter2', logRounds: 1)),
        throwsArgumentError,
      );
      expect(
        () => hasher.execute(const BcryptHashInput(plaintext: 'hunter2', logRounds: 99)),
        throwsArgumentError,
      );
    });
  });

  group('BcryptVerifier', () {
    test('accepts the correct plaintext against a freshly generated hash', () {
      final hash = hasher.execute(const BcryptHashInput(plaintext: 'correct horse battery staple', logRounds: 4));
      final result = verifier.execute(
        BcryptVerifyInput(plaintext: 'correct horse battery staple', hash: hash.hash),
      );
      expect(result.matches, isTrue);
    });

    test('rejects the wrong plaintext against a valid hash', () {
      final hash = hasher.execute(const BcryptHashInput(plaintext: 'correct horse battery staple', logRounds: 4));
      final result = verifier.execute(
        BcryptVerifyInput(plaintext: 'wrong password', hash: hash.hash),
      );
      expect(result.matches, isFalse);
    });

    test('verifies against a pre-recorded fixture hash (fixed-salt regression vector)', () {
      // bcrypt embeds a random salt in every hash, so there is no
      // cross-implementation "hash of 'abc'" constant the way there is for
      // MD5/SHA-256. Instead this fixture was generated once with the
      // `bcrypt` package itself using a deterministic (seeded, non-secure)
      // PRNG for the salt, then frozen here so the verifier's decoding of an
      // externally-produced hash string is exercised, not just a live
      // hash-then-immediately-verify round trip against fresh output.
      const fixtureHash = r'$2a$04$y4tiGeWwdpFlQhdt/zqxf.ZbJzRxB2WjyEZXYjWxuSEEsplrY8JQ2';

      final correct = verifier.execute(
        const BcryptVerifyInput(plaintext: 'correct horse battery staple', hash: fixtureHash),
      );
      expect(correct.matches, isTrue);

      final wrong = verifier.execute(
        const BcryptVerifyInput(plaintext: 'wrong password', hash: fixtureHash),
      );
      expect(wrong.matches, isFalse);
    });

    test('returns false (not a thrown error) for a malformed hash string', () {
      final result = verifier.execute(
        const BcryptVerifyInput(plaintext: 'hunter2', hash: 'not-a-bcrypt-hash'),
      );
      expect(result.matches, isFalse);
    });
  });
}
